src/aio.rs

```diff
@@ -577,3 +577,111 @@ impl ConnectionLike for MultiplexedConnection {
         self.db
     }
 }
+
+#[cfg(feature = "connection-manager")]
+mod connection_manager {
+    use super::*;
+
+    use std::sync::Arc;
+
+    use arc_swap::ArcSwap;
+
+    /// A `ConnectionManager` is a proxy that wraps a multiplexed connection and
+    /// automatically reconnects to the server when necessary.
+    #[derive(Clone)]
+    pub struct ConnectionManager {
+        connection_info: ConnectionInfo,
+        connection: ArcSwap<MultiplexedConnection>,
+    }
+
+    impl ConnectionManager {
+        /// Connect to the server and store the connection inside the returned `ConnectionManager`.
+        ///
+        /// This requires the `connection-manager` feature, which will also pull in
+        /// the Tokio executor.
+        pub async fn new(connection_info: ConnectionInfo) -> RedisResult<Self> {
+            // Wait for the initial connection to be established
+            let inner_connection = connect(&connection_info).await?;
+
+            // Wrap the connection in a MultiplexedConnection
+            let (connection, driver) = MultiplexedConnection::new(inner_connection);
+
+            // Spawn the driver that drives the connection future
+            tokio::spawn(driver);
+
+            // Wrap the connection in an `ArcSwap` instance for fast atomic access
+            Ok(Self {
+                connection_info,
+                connection: ArcSwap::from(Arc::new(connection)),
+            })
+        }
+
+        async fn reconnect(&mut self) -> RedisResult<&mut Self> {
+            // TODO: While this code uses compare-and-swap to prevent active
+            // connections from being overwritten in the ConnectionManager
+            // struct, it can still lead to multiple connections being
+            // established in parallel.
+            let guard = self.connection.load();
+
+            // Establish a new connection
+            println!("Connect");
+            let (new_connection, driver) = MultiplexedConnection::new(
+                connect(&self.connection_info).await?
+            );
+
+            // Spawn the driver that drives the connection future
+            tokio::spawn(driver);
+
+            // Update the connection in the connection manager
+            self.connection.compare_and_swap(guard, Arc::new(new_connection));
```

Comment:
@Marwes this is not sufficient to avoid creating multiple connections in parallel, right? The way I understand things, it will only prevent *storing* multiple connections (overwriting the previously established connection). In a multi-threaded scenario, the `.await` call on the `connect()` method may yield control to another task that also launches a connection.

If that is correct, I will add an atomic connection counter to avoid reconnecting multiple times.

I didn't manage to trigger multiple connections when removing the loop delay in the example, but that's probably just because I did not enqueue reconnection tasks from multiple threads at the same time.

Reply:
We can't prevent creating the future twice, but it is possible to prevent the reconnection from happening twice. Something like this.

```rust
// We can reuse the guard/connection reference that were used to run the connection
fn reconnect(&self,  current: arc_swap::Guard<_>) {
    let (new_connection, driver) = MultiplexedConnection::new(
        connect(&self.connection_info).await?
    );

    // Update the connection in the connection manager
    let prev = self.connection.compare_and_swap(guard, Arc::new(new_connection));

    if Arc::ptr_eq(&prev, current) {
        tokio::spawn(driver);
        tokio::spawn(new_connection);
    }
}
```

Reply:
Hmm, but `connect(...).await` will already open the TCP socket, right?

Reply:
Also, `new_connection` is a pipeline. Does it need to be spawned? I thought only the driver needs to be spawned.

Reply:
```
    let new_connection =  async move {
        let (new_connection, driver) = MultiplexedConnection::new(
            connect(&self.connection_info).await?
        );
        tokio::spawn(driver);
        Ok(new_connection)
    }.shared();

    // Update the connection in the connection manager
    let prev = self.connection.compare_and_swap(guard, Arc::new(new_connection));

    if Arc::ptr_eq(&prev, current) {
        // start the connection attempt immediately but do not wait on it
        tokio::spawn(new_connection);
    }
```

Reply:
Ok, that makes sense. However, now the `ArcSwap` doesn't wrap a `MultiplexedConnection` anymore, but a `dyn Future<MultiplexedConnection>` instead: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6ecee2780e2046452967f70fa97d5f3e

I tried putting a `Shared` in the `ArcSwap` instead. However, with static dispatch this does of course not work due to a type change in the future. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=36e7e78a91db04a403c06ec379c565e6

I tried to use dynamic dispatch instead but couldn't get it to work :confused: I'll try again tomorrow, but if you have any pointer that would be helpful. (These are the situations where I sometimes still bump into the Rust async APIs, even though I've done async stuff for a while now.)

Reply:
You will indeed need to box it and use dynamic dispatch https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=99941c415672f38d43b81c7b7cb67b3b

Reply:
Great, thanks! Somehow it did not occur to me to put the pinned box *inside* the Shared instance.

I also managed to simplify the code a bit. `boxed()` already returns a pinned box, so the type annotations and `Box::new` calls are not necessary. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b62964c32866c8ab11af0912eff1d4c9

Reply:
Hmm, it seems that since reconnecting can fail, I'll have to wrap the `RedisError` in an `Arc` since it must be `Clone` to be put inside `Shared`...

Reply:
@Marwes I hit another blocker: When a reconnection starts, the future is stored. All commands that are issued after that point in time should await the connection future before they can issue the command.

The problem is that `Shared` cannot contain a `Future` that resolves to a `RedisError` because that is not `Clone`. I solved this by using `Result<T, Arc<RedisError>>` instead, which works generally but leads to the following issue: If `req_packed_command` fails, it must return a `RedisError`. From the reconnection we get an `Arc<RedisError>` but cannot clone the error (since it's not `Clone`).

I see three possibilities:

- Put `io::Error` inside `RedisError` into something like an `Arc` so that it can be cloned, or use a cloneable abstraction [like this one (seems unmaintained)](https://crates.io/crates/arc-io-error) instead. However, that would be a breaking change.
- Convert `ErrorRepr::IoError(io::Error)` into `ErrorRepr::WithDescription(ErrorKind::IoError, "Reconnecting failed")` and discard the error details. (They should probably be logged though, see #277.) Main downside of this is that errors returned by initial connection and reconnection are different (since error details are discarded). The user does not see *why* the reconnection failed (except in logs). Note that a failed reconnection could immediately trigger another attempt.
- Change the implementation logic. As long as the connection is not yet established, fail all commands immediately with `ErrorRepr::WithDescription(ErrorKind::IoError, "Connection lost, reconnection in progress") until the connection is re-established.

I tend towards number 2. What do you think?

Reply:
I implemented number 2 for now. Seems to work! We can also pass in a string representation of the I/O error to the cloned error.

Reply:
I think ideally we would return the original `io::Error` to the first one to see the error, while the rest get a generic `Reconnecting failed` one. May need to use a `Mutex` for that though, unsure if there is a better way than that :/